### PR TITLE
Allow choosing different background/audio files for individual difficulties

### LINKED ISSUE #30806: Adding a new background/audio file in the editor may remove file which is used in other difficulties
The method for replacing the background doesn't currently check if the old background was used or not:

https://github.com/ppy/osu/blob/7cd724f342adca027eee034ce352a1a4dfc1a458/osu.Game/Screens/Edit/Setup/ResourcesSection.cs#L84-L94

Same applies to audio file.

### Discussed in https://github.com/ppy/osu/discussions/30435

<div type='discussions-op-text'>

<sup>Originally posted by **AFasterSlowpoke** October 27, 2024</sup>
When I drag an image to the osu lazer editor to set the background, it changes the background for all difficulties, instead of just the difficulty i'm currently editing.

I think that it would improve the osu lazer editor if dragging and dropping a new background only changed the current difficulty's background, with a button to set the current difficulty's background to all other difficulties.</div>

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes https://github.com/ppy/osu/issues/30806

RFC. I was going to start by following what the title of the issue thread states, which is to just not delete files that are used by other difficulties. However, the underlying discussion thread asks for more than that, and such a change does not really make much sense if the general code still assumes all beatmaps should use one background/audio file, so I took it one step further and changed the whole logic to properly allow difficulties to have their own background/audio.

Preview:

https://github.com/user-attachments/assets/75e47c16-52cc-45af-8d4f-bd05dffb08d9

### CODE REVIEW

**Comment by Phippe** [comment]
I would've assumed that there would be something like a default image for all difficulties. Then each difficulty would have the option to overwrite the default image, but also to change back to it.

This would make the case easier where there are a large number of difficulties, but only a few of them have custom backgrounds. If you wanted to change the background for all the difficulties that share the same default background, you would have to do so for each difficulty separately.

Just my two cents though, I'm not a mapper.

**Comment by peppy** [comment]
Hmm, I dunno. I thought this PR is providing a pretty good UX until we actually have a concept of set level metadata. It meets existing users' expectations.

Adding a "this beatmap overrides the default" is going to be a pain to implement a good UX for. But I guess I'm all ears?

> This would make the case easier where there are a large number of difficulties, but only a few of them have custom backgrounds. If you wanted to change the background for all the difficulties that share the same default background, you would have to do so for each difficulty separately.

This seems *very* edge case.

**Comment by bdach** [comment]
Easiest would be to ask user on file selection whether to use it on all difficulties or just the one.

**Comment by peppy** [comment]
Kinda the same thing really, just more explicit.

But it would solve my remaining issue with this PR, which is that it is applying the action to both background and audio at the same time (should really be per resource).

**Comment by frenzibyte** [comment]
> But it would solve my remaining issue with this PR, which is that it is applying the action to both background and audio at the same time (should really be per resource).

It is technically doing it per resource if you check the code. Is the frontend not conveying that well?

> > I would've assumed that there would be something like a default image for all difficulties. Then each difficulty would have the option to overwrite the default image, but also to change back to it.
> 
> I would also assume so yes.
> 
> This looks half-baked to me. There should be a way to replace the background on all difficulties at once, and it should be the default. Replacing on a single difficulty should be the exception. And maybe there should be a dropdown which allows picking from existing backgrounds from all the difficulties of the set.

The way it was implemented in this PR followed the natural storage of metadata in beatmaps, choosing a background/audio updates the metadata of the current difficulty, and the button changes all other difficulties to update as such.

I've also thought about a dialog after opening the PR but I wasn't sure if it's worth the effort. Now that it's brought up, and the button is already misleading in the idea of "does it update both background and audio or is it per-resource", so I will update the PR as such.

**Comment by peppy** [comment]
@frenzibyte to confirm we're all on the same page, I think spaceman is proposing that in the existing file chooser popup, after selecting a file, the file chooser content is replaced with two buttons (in the same popup) which let the mapper select [apply to this difficulty] [apply to all difficulties]. Should only show this if there's more than one difficulty already.

**Comment by frenzibyte** [comment]
OP updated with preview of the new behaviour.

File: osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            : base(handledExtensions)
+        {
+            this.beatmapHasMultipleDifficulties = beatmapHasMultipleDifficulties;
+        }
+
+        protected override FileChooserPopover CreatePopover(string[] handledExtensions, Bindable<FileInfo?> current, string? chooserPath)
+        {
+            var popover = new BeatmapFileChooserPopover(handledExtensions, current, chooserPath, beatmapHasMultipleDifficulties);
+
+            popover.ApplyToAllDifficulties.ValueChanged += v =>
```
Comment (peppy) [inline osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs:44]: What is this flow? Why isn't this a `BindTo` or something more standard?

File: osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                if (beatmapHasMultipleDifficulties)
+                    selectApplicationScopeContainer.FadeIn(200, Easing.InQuint);
+                else
+                    base.OnFileSelected(file);
+            }
+
+            private void onChangeScopeSelected(ValueChangedEvent<bool?> c)
+            {
+                if (c.NewValue == null)
+                    return;
```
Comment (peppy) [inline osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs:154]: How can this happen?

File: osu.Game/Screens/Edit/Setup/ResourcesSection.cs [hunk 3]
```diff
@@ -72,59 +77,115 @@ private void load()
             audioTrackChooser.Current.BindValueChanged(audioTrackChanged);
         }
 
-        public bool ChangeBackgroundImage(FileInfo source)
+        public bool ChangeBackgroundImage(FileInfo source, bool applyToAllDifficulties)
```
Comment (peppy) [inline osu.Game/Screens/Edit/Setup/ResourcesSection.cs:80]: I get that the logic was split out in `master`, but now that this is 100x more complex I'd hope that the complex part was shared between both implementations. Is there a reason you didn't do this while writing it?

File: osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Screens/Edit/Setup/FormBeatmapFileSelector.cs]: The idea was that the popover's bindable is nullable so that whatever the user chooses will always trigger a change in the bindable, but I don't want the general bindable to be nullable as well so there's a value changed event here.

Anyways I can understand why it looks odd, I've rewritten it in https://github.com/ppy/osu/pull/30860/commits/4a1401a33df7c3b489c6c0db05b68f6f1fe31079

File: osu.Game/Screens/Edit/Setup/ResourcesSection.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Screens/Edit/Setup/ResourcesSection.cs:80]: I felt this is still within the "it's fine" area, and that if I tried to make it shared then I'll end up with an extremely weird method that begs for itself to be inlined.

Diff will be something like this:
```diff
diff --git a/osu.Game/Screens/Edit/Setup/ResourcesSection.cs b/osu.Game/Screens/Edit/Setup/ResourcesSection.cs
index 1ce944b5a4..a02900a204 100644
--- a/osu.Game/Screens/Edit/Setup/ResourcesSection.cs
+++ b/osu.Game/Screens/Edit/Setup/ResourcesSection.cs
@@ -82,57 +82,12 @@ public bool ChangeBackgroundImage(FileInfo source, bool applyToAllDifficulties)
             if (!source.Exists)
                 return false;
 
-            var set = working.Value.BeatmapSetInfo;
-
-            if (applyToAllDifficulties)
-            {
-                string newFilename = $@"bg{source.Extension}";
-
-                foreach (var beatmapInSet in set.Beatmaps)
-                {
-                    if (set.GetFile(beatmapInSet.Metadata.BackgroundFile) is RealmNamedFileUsage existingFile)
-                        beatmaps.DeleteFile(set, existingFile);
-
-                    if (beatmapInSet.Metadata.BackgroundFile != newFilename)
-                    {
-                        beatmapInSet.Metadata.BackgroundFile = newFilename;
-
-                        if (!beatmapInSet.Equals(working.Value.BeatmapInfo))
-                            beatmaps.Save(beatmapInSet, beatmaps.GetWorkingBeatmap(beatmapInSet).Beatmap);
-                    }
-                }
-            }
-            else
-            {
-                var beatmap = working.Value.BeatmapInfo;
-
-                string[] filenames = set.Files.Select(f => f.Filename).Where(f =>
-                    f.StartsWith(@"bg", StringComparison.OrdinalIgnoreCase) &&
-                    f.EndsWith(source.Extension, StringComparison.OrdinalIgnoreCase)).ToArray();
-
-                string currentFilename = working.Value.Metadata.BackgroundFile;
-
-                var oldFile = set.GetFile(currentFilename);
-                string? newFilename = null;
-
-                if (oldFile != null && set.Beatmaps.Where(b => !b.Equals(beatmap)).All(b => b.Metadata.BackgroundFile != currentFilename))
-                {
-                    beatmaps.DeleteFile(set, oldFile);
-                    newFilename = currentFilename;
-                }
-
-                newFilename ??= NamingUtils.GetNextBestFilename(filenames, $@"bg{source.Extension}");
-                working.Value.Metadata.BackgroundFile = newFilename;
-            }
-
-            using (var stream = source.OpenRead())
-                beatmaps.AddFile(set, stream, working.Value.Metadata.BackgroundFile);
-
-            editorBeatmap.SaveState();
+            changeResource(source, applyToAllDifficulties, @"bg",
+                metadata => metadata.BackgroundFile,
+                (metadata, name) => metadata.BackgroundFile = name);
 
             headerBackground.UpdateBackground();
             editor?.ApplyToBackground(bg => bg.RefreshBackground());
-
             return true;
         }
 
@@ -141,20 +96,34 @@ public bool ChangeAudioTrack(FileInfo source, bool applyToAllDifficulties)
             if (!source.Exists)
                 return false;
 
+            changeResource(source, applyToAllDifficulties, @"audio",
+                metadata => metadata.AudioFile,
+                (metadata, name) => metadata.AudioFile = name);
+
+            music.ReloadCurrentTrack();
+            return true;
+        }
+
+        private void changeResource(FileInfo source, bool applyToAllDifficulties, string baseFilename, Func<BeatmapMetadata, string> readFilename, Action<BeatmapMetadata, string> writeFilename)
+        {
             var set = working.Value.BeatmapSetInfo;
 
+            string newFilename = string.Empty;
+
             if (applyToAllDifficulties)
             {
-                string newFilename = $@"audio{source.Extension}";
+                newFilename = $"{baseFilename}{source.Extension}";
 
                 foreach (var beatmapInSet in set.Beatmaps)
                 {
-                    if (set.GetFile(beatmapInSet.Metadata.AudioFile) is RealmNamedFileUsage existingFile)
+                    string filenameInBeatmap = readFilename(beatmapInSet.Metadata);
+
+                    if (set.GetFile(filenameInBeatmap) is RealmNamedFileUsage existingFile)
                         beatmaps.DeleteFile(set, existingFile);
 
-                    if (beatmapInSet.Metadata.AudioFile != newFilename)
+                    if (filenameInBeatmap != newFilename)
                     {
-                        beatmapInSet.Metadata.AudioFile = newFilename;
+                        writeFilename(beatmapInSet.Metadata, newFilename);
 
                         if (!beatmapInSet.Equals(working.Value.BeatmapInfo))
                             beatmaps.Save(beatmapInSet, beatmaps.GetWorkingBeatmap(beatmapInSet).Beatmap);
@@ -166,31 +135,29 @@ public bool ChangeAudioTrack(FileInfo source, bool applyToAllDifficulties)
                 var beatmap = working.Value.BeatmapInfo;
 
                 string[] filenames = set.Files.Select(f => f.Filename).Where(f =>
-                    f.StartsWith(@"audio", StringComparison.OrdinalIgnoreCase) &&
+                    f.StartsWith(baseFilename, StringComparison.OrdinalIgnoreCase) &&
                     f.EndsWith(source.Extension, StringComparison.OrdinalIgnoreCase)).ToArray();
 
-                string currentFilename = working.Value.Metadata.AudioFile;
+                string currentFilename = readFilename(working.Value.Metadata);
 
                 var oldFile = set.GetFile(currentFilename);
-                string? newFilename = null;
 
-                if (oldFile != null && set.Beatmaps.Where(b => !b.Equals(beatmap)).All(b => b.Metadata.AudioFile != currentFilename))
+                if (oldFile != null && set.Beatmaps.Where(b => !b.Equals(beatmap)).All(b => readFilename(b.Metadata) != currentFilename))
                 {
                     beatmaps.DeleteFile(set, oldFile);
                     newFilename = currentFilename;
                 }
 
-                newFilename ??= NamingUtils.GetNextBestFilename(filenames, $@"audio{source.Extension}");
-                working.Value.Metadata.AudioFile = newFilename;
+                if (string.IsNullOrEmpty(newFilename))
+                    newFilename = NamingUtils.GetNextBestFilename(filenames, $@"{baseFilename}{source.Extension}");
+
+                writeFilename(working.Value.Metadata, newFilename);
             }
 
             using (var stream = source.OpenRead())
-                beatmaps.AddFile(set, stream, working.Value.Metadata.AudioFile);
+                beatmaps.AddFile(set, stream, newFilename);
 
             editorBeatmap.SaveState();
-            music.ReloadCurrentTrack();
-
-            return true;
         }
 
         private void backgroundChanged(ValueChangedEvent<FileInfo?> file)
```

I'm shaving a good amount of code duplication so I personally wouldn't mind pushing it in, but I just want to make sure this doesn't look over-engineered to you.

Comment (peppy) [reply osu.Game/Screens/Edit/Setup/ResourcesSection.cs:80]: I think that's fine as a starting point, I'll review after you push.

**Comment by frenzibyte** [comment]
In addition to the above comments, I've also looked back to how I was ensuring changes to metadata of other difficulties is being saved and I realised my logic was unnecessarily complicated.

Instead, I've updated the metadata of the other difficulties from `beatmap.Value.BeatmapSetInfo` (which is the detached beatmap set of the current difficulty), and I've forced a `Save` call after changing each resource for two reasons:
 1. `EditorChangeHandler` cannot be aware of changes to metadata of other difficulties, therefore if the user changes resource on a specific difficulty, then switches to another difficulty without saving, the user will not be prompted to save, and when they change the resource on the other difficulty, the state of the previous difficulty will be stale.
 2. If the user presses "Forget all changes" on a save dialog after changing a resource, then the difficulty/beatmap will become corrupted if the metadata was pointing to an old file that has been removed by the change logic.

Changing resources should be a rare process anyway, so I don't think triggering a save will make anybody sad.

File: osu.Game/Screens/Edit/Setup/ResourcesSection.cs [hunk 4]
```diff
@@ -72,75 +74,92 @@ private void load()
...
 
-            using (var stream = source.OpenRead())
+                foreach (var beatmap in set.Beatmaps)
+                {
+                    if (set.GetFile(readFilename(beatmap.Metadata)) is RealmNamedFileUsage otherExistingFile)
+                        beatmaps.DeleteFile(set, otherExistingFile);
+
+                    writeFilename(beatmap.Metadata, newFilename);
+                }
+            }
```
Comment (bdach) [inline osu.Game/Screens/Edit/Setup/ResourcesSection.cs:121]: This logic is missing a completely key step.

Recall that a beatmap's background filename is contained in the `.osu` file for the beatmap. None of this actually mutates the other `.osu` files. Which means that this breaks immediately on export and re-import; only the difficulty that was active at the point of replacing all backgrounds will have the backgrounds, and the other difficulties will point to dead assets that are no longer there because they were deleted at the point of replace. (I have tested this.)

In other words this logic needs to _re-encode every single beatmap in the set_ to work correctly.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Edit/Setup/ResourcesSection.cs [hunk 5]
```diff
@@ -46,14 +46,16 @@ private void load()
...
             Children = new Drawable[]
             {
-                backgroundChooser = new FormFileSelector(".jpg", ".jpeg", ".png")
+                backgroundChooser = new FormBeatmapFileSelector(beatmapHasMultipleDifficulties, ".jpg", ".jpeg", ".png")
                 {
                     Caption = GameplaySettingsStrings.BackgroundHeader,
                     PlaceholderText = EditorSetupStrings.ClickToSelectBackground,
                 },
-                audioTrackChooser = new FormFileSelector(".mp3", ".ogg")
+                audioTrackChooser = new FormBeatmapFileSelector(beatmapHasMultipleDifficulties, ".mp3", ".ogg")
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Edit/Setup/ResourcesSection.cs:58]: I'm still of the thought that we should leave such things up to the ranking criteria. I'd rather have a defined flow for this, because if we don't, users will end up documenting weird and potentially incorrect ways of doing it.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Edit/Setup/ResourcesSection.cs:58]: Well in that case the same thing that applies in https://github.com/ppy/osu/pull/30860#discussion_r1865569965 applies here, because the audio file is *also* encoded to `.osu`.

File: osu.Game/Screens/Edit/Setup/ResourcesSection.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Screens/Edit/Setup/ResourcesSection.cs]: Should be addressed in https://github.com/ppy/osu/pull/30860/commits/8e0f6fc12dc04a224a9aefb0121f990a8b007af2, I don't like the part where I'm re-fetching each difficulty's skin in order to preserve it but can't see any other way around that.

**Comment by peppy** [comment]
I rewrote the `changeResource` method to be better to my eye. @bdach would appreciate a one-over (and also @frenzibyte to check it's not worse than before).

The main goal here was that I wanted to avoid doing the per-difficulty save in the case where the filename didn't change. In testing, it can take a few seconds to run this as it triggers a full difficulty calculation via 

https://github.com/ppy/osu/blob/5cb6b86b1cd0814b76fcd11cfcf29a04680d4022/osu.Game/Beatmaps/BeatmapManager.cs#L49

I'm probably okay with this change now. We can revisit making the save more optimal in the future.

**Review by peppy — APPROVED** [review]

**Review by frenzibyte — COMMENTED** [review]
> The main goal here was that I wanted to avoid doing the per-difficulty save in the case where the filename didn't change. 

~~That should've been already correctly handled in my previous code, I've always checked if the filename of the other diff is different first before writing to it and triggering a save. If anything, your code is just emphasizing that part, which is good.~~ I take this back, everything is good.

I'm not 100% sure how to feel about there being two small `applyToAllDifficulties` if conditionals with some shared code placed between them. It feels slightly awkward to me when reading the method from top to bottom but I can understand why it's done this way (to make the method look more like it's working step-by-step but each step behaves differently based on the condition). I can live with it.

**Comment by peppy** [comment]
> That should've been already correctly handled in my previous

Check the older commit, it wasn't. I know because i debugged it and that's why I refactored everything.

**Review by bdach — APPROVED** [review]

**Comment by bdach** [comment]
Seems to behave fine in brief testing, soooo
